hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/expression/json/JsonPathUtil.java

```diff
@@ -106,4 +110,78 @@ public static String serialize(Object object) {
             throw new RuntimeException(e);
         }
     }
+
+    /**
+     * Implementation of fixed-capacity cache based on {@link ConcurrentHashMap}
+     * caching the initial set of keys.
+     * <p>
+     * The cache has no eviction policy, once an element is put into it, it stays
+     * there so long as the cache exists. Once the cache is full, no new items are
+     * cached.
+     * <p>
+     * It's designed for caching of compiled JSONPath expressions in the context of
+     * one query execution, based on the assumption that typically there's a low
+     * number of distinct expressions that fit into the cache and that the
+     * expressions come in arbitrary order. If there number of distinct expressions
+     * is larger than capacity, we assume that some are more common than others, and
+     * we're likely to observe those at the beginning and cache those. If the number
+     * of expressions exceeds the capacity many times, we'll cache arbitrary few of
+     * them and the rest will be calculated each time without caching - a similar
+     * behavior to what an LRU cache will provide, but without the overhead of usage
+     * tracking. Degenerate case is when items are sorted by the cache key - after
+     * the initial phase the cache will have zero hit rate.
+     * <p>
+     * Note: The size of the inner map may become bigger than maxCapacity if there
+     * are multiple concurrent computeIfAbsent executions. We don't address this for
+     * the purpose of optimizing the read performance. The amount the size can
+     * exceed the limit is bounded by the number of concurrent writers.
+     */
+    public static class ConcurrentInitialSetCache<K, V> implements IdentifiedDataSerializable, Serializable {
```

Comment:
We don't need to serialize this class. If I'm not mistaken, at the time we serialize it, it's always empty. It should be enough to mark the `pathCache` fields as transient, and perhaps initialize it in `JsonQueryFunction.readObject()`, because transient fields aren't assigned with the field initializer after deserialization.

Reply:
I did a lazy initialization during first ```eval```. I think this may be even better, since you won't calculate the path before submission of the job.